### PR TITLE
feat(now): implement `now()` in batch

### PR DESCRIPTION
**Pull request by soundOfDestiny**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Enable batch system to evaluate `NOW()` as a function instead of an operator.
result before this PR:
```
dev=> select now();
ERROR:  QueryError: internal error: Failed to build executor: Expr error: Unsupported function: Ok(Now)
```
result after this PR:
test result:
```
dev=> select now();
          ?column?          
----------------------------
 2023-01-03 08:01:52.304273
(1 row)

dev=> create table t (v1 int);
CREATE_TABLE
dev=> create materialized view mv as select now() as nw from t where v1 >= 3;
ERROR:  QueryError: internal error: Rpc error: gRPC error (Internal error): Expression error: Expected session timestamp bound into Now
dev=> create table tt(v1 int, t1 timestamp); 
CREATE_TABLE
dev=> create materialized view mv as select * from tt where tt.t1 > now() - INTERVAL '1 hour';
CREATE_MATERIALIZED_VIEW
```
We insert a session timestamp into arguments of `NOW()`, which will only be used in batch system.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#7139

### CODE REVIEW

**Comment by fuyufjh** [comment]
```
dev=> select now();
ERROR:  QueryError: internal error: Failed to build executor: Expr error: Unsupported function: Ok(Now)
```

Is it intended?

File: src/expr/src/expr/mod.rs [hunk 1]
```diff
@@ -148,6 +150,15 @@ pub fn build_from_prost(prost: &ExprNode) -> Result<BoxedExpression> {
             ArrayConcatExpression::try_from(prost).map(Expression::boxed)
         }
         Vnode => VnodeExpression::try_from(prost).map(Expression::boxed),
+        Now => {
```
Comment (jon-chuang) [inline src/expr/src/expr/mod.rs:153]: A bit afraid this change will affect use of now in streaming. 

Currently, I think the frontend will Err if the now appears in an invalid location. However, this change may cause the now in stream to evaluate to the static session_timestamp value.

File: src/frontend/src/binder/expr/function.rs [hunk 2]
```diff
@@ -290,7 +290,13 @@ impl Binder {
             // TODO: choose which pg version we should return.
             "version" => return Ok(ExprImpl::literal_varchar("PostgreSQL 13.9-RW".to_string())),
             // non-deterministic
-            "now" => ExprType::Now,
+            "now" => {
```
Comment (jon-chuang) [inline src/frontend/src/binder/expr/function.rs:293]: Perhaps one solution is to only bind the session_timestamp if the session is batch?

File: src/expr/src/expr/mod.rs (hunk 1 above)
Comment (jon-chuang) [reply src/expr/src/expr/mod.rs:153]: Actually planner tests don't help here as the planner will still evaluate it to `Now` expr.

**Comment by soundOfDestiny** [comment]
> In the PR description:
> 
> ```
> dev=> select now();
> ERROR:  QueryError: internal error: Failed to build executor: Expr error: Unsupported function: Ok(Now)
> ```
> 
> Is it intended?

PR description updated.
That should be result before this PR.

File: src/frontend/src/optimizer/plan_node/logical_project.rs [hunk 3]
```diff
@@ -305,6 +306,14 @@ impl ToStream for LogicalProject {
         let (input, input_col_change) = self.input().logical_rewrite_for_stream(ctx)?;
         let (proj, out_col_change) = self.rewrite_with_input(input.clone(), input_col_change);
 
+        if proj.exprs().iter().any(|expr| expr.count_nows() > 0) {
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_project.rs:309]: I think this is quite hacky? Now can be placed in almost any location including order by expression. The way we handle it currently is not so good. Perhaps we can try to ban binding of now if it is not in a where clause?

Comment (soundOfDestiny) [reply src/frontend/src/optimizer/plan_node/logical_project.rs:309]: > I think this is quite hacky? Now can be placed in almost any location including order by expression. The way we handle it currently is not so good. Perhaps we can try to ban binding of now if it is not in a where clause?

but batch needs to bind `now` outside where clause..

File: src/expr/src/expr/mod.rs (hunk 1 above)
Comment (soundOfDestiny) [reply src/expr/src/expr/mod.rs:153]: > A bit afraid this change will affect use of now in streaming.
> 
> Currently, I think the frontend will Err if the now appears in an invalid location. However, this change may cause the now in stream to evaluate to the static `session_timestamp` value. I don't think this is intended.

I have just added some check and a planner test for it.

File: src/frontend/src/binder/expr/function.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/frontend/src/binder/expr/function.rs:293]: > Perhaps one solution is to only bind the session_timestamp if the session is batch?

how?

File: src/frontend/src/optimizer/plan_node/logical_project.rs (hunk 3 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_node/logical_project.rs:309]: Yes, seems there are two solutions:
1. Bind now as valid expression only in batch
2. Binder will reject now in any location but WHERE clause (i.e. filter) when creating stream

Both require `Binder` to be aware of whether it is binding from batch or stream.

File: src/frontend/src/binder/expr/function.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/frontend/src/binder/expr/function.rs:293]: > Perhaps one solution is to only bind the session_timestamp if the session is batch?
> 
> ~session~ -> query

fixed

File: src/frontend/src/optimizer/plan_node/logical_project.rs (hunk 3 above)
Comment (soundOfDestiny) [reply src/frontend/src/optimizer/plan_node/logical_project.rs:309]: > Yes, seems there are two solutions:
> 
> 1. Bind now as valid expression with the `session_timestamp` input only in batch
> 2. Binder will reject now in any location but WHERE clause (i.e. filter) when creating stream
> 
> Both require `Binder` to be aware of whether it is binding from batch or stream. But this is trivial, we can have e.g. `Binder::new_for_stream` when handling create MV.
> 
> I prefer the latter solution as we do not have to wait for the compute node to try to create an executor for it to fail.We already intercept the `now` expression in `WHERE` clauses...
> 
> Actually, I prefer if _both_ solutions are implemented. This way, `now` should have arity 1 in batch mode and arity 0 in stream mode. Why not you work on the 1. and I can PR in something for 2.

thanks. I have just done 1.

**Comment by soundOfDestiny** [comment]
test result:
```
dev=> select now();
          ?column?          
----------------------------
 2023-01-03 08:01:52.304273
(1 row)

dev=> create table t (v1 int);
CREATE_TABLE
dev=> create materialized view mv as select now() as nw from t where v1 >= 3;
ERROR:  QueryError: internal error: Rpc error: gRPC error (Internal error): Expression error: Expected session timestamp bound into Now
dev=> create table tt(v1 int, t1 timestamp); 
CREATE_TABLE
dev=> create materialized view mv as select * from tt where tt.t1 > now() - INTERVAL '1 hour';
CREATE_MATERIALIZED_VIEW
```

File: src/frontend/src/binder/mod.rs [hunk 4]
```diff
@@ -77,24 +80,40 @@ pub struct Binder {
...
 
     search_path: SearchPath,
+
+    in_create_mv: bool,
 }
 
 impl Binder {
-    pub fn new(session: &SessionImpl) -> Binder {
+    fn new_inner(session: &SessionImpl, in_create_mv: bool) -> Binder {
+        let now_nsecs = Local::now().timestamp_nanos();
```
Comment (fuyufjh) [inline src/frontend/src/binder/mod.rs:89]: To keep the result of batch queries consistent with streaming, it's better to use the timestamp of the current epoch. (had mentioned in https://github.com/risingwavelabs/risingwave/issues/7139#issuecomment-1367734075)

**Comment by chenzl25** [comment]
Once we intend to support `now()`, we are better to consider it with the timezone. PostgreSQL has a session variable `TimeZone` to allow users to set the time zone for displaying and interpreting time stamps.

```
tpch=# select now();
              now
-------------------------------
 2023-01-03 16:24:21.178595+08
(1 row)

tpch=# set timezone = 'UTC';
SET
tpch=# select now();
              now
-------------------------------
 2023-01-03 08:24:48.025885+00
(1 row)
```

File: src/frontend/src/binder/mod.rs (hunk 4 above)
Comment (soundOfDestiny) [reply src/frontend/src/binder/mod.rs]: > se the timestamp of the current epoch

sry. fixed.

File: src/expr/src/expr/mod.rs [hunk 5]
```diff
@@ -148,6 +150,16 @@ pub fn build_from_prost(prost: &ExprNode) -> Result<BoxedExpression> {
...
+            let rex_node = try_match_expand!(prost.get_rex_node(), Ok)?;
+            let RexNode::FuncCall(func_call_node) = rex_node else {
+                bail!("Expected RexNode::FuncCall in Now");
+            };
+            let Some(session_timestamp) = func_call_node.children.first() else {
```
Comment (jon-chuang) [inline src/expr/src/expr/mod.rs:158]: update the var names?

File: src/frontend/planner_test/tests/testdata/expr.yaml [hunk 6]
```diff
@@ -380,3 +380,17 @@
     create table t (v1 timestamp, v2 timestamp);
     select * from t where v1 >= now() or v2 >= now();
   stream_error: 'Expr error: Invalid parameter now: now expression must be placed in a comparison'
+- name: now expression condition without filter
```
Comment (jon-chuang) [inline src/frontend/planner_test/tests/testdata/expr.yaml:383]: Can remove this case, it is handled in https://github.com/risingwavelabs/risingwave/pull/7167

**Comment by soundOfDestiny** [comment]
> Once we intend to support `now()`, we are better to consider it with the timezone. PostgreSQL has a session variable `TimeZone` to allow users to set the time zone for displaying and interpreting time stamps.
> 
> ```
> tpch=# select now();
>               now
> -------------------------------
>  2023-01-03 16:24:21.178595+08
> (1 row)
> 
> tpch=# set timezone = 'UTC';
> SET
> tpch=# select now();
>               now
> -------------------------------
>  2023-01-03 08:24:48.025885+00
> (1 row)
> ```

Should we use `Timestamptz`?

**Comment by jon-chuang** [comment]
> Once we intend to support now(), we are better to consider it with the timezone.

Hmm, will the epoch be aware of this timezone (i.e. it is configured to start from UTC 0?)

Should the stream also be aware of this timezone? How does cmp between timestamp with timezone and timestamp work?

File: src/expr/src/expr/mod.rs (hunk 5 above)
Comment (soundOfDestiny) [reply src/expr/src/expr/mod.rs]: > update the var names?

fixed

File: src/frontend/planner_test/tests/testdata/expr.yaml (hunk 6 above)
Comment (soundOfDestiny) [reply src/frontend/planner_test/tests/testdata/expr.yaml]: > Can remove this case, it is handled in #7167

fixed

**Comment by chenzl25** [comment]
> > Once we intend to support now(), we are better to consider it with the timezone.
> 
> Hmm, will the epoch be aware of this timezone (i.e. it is configured to start from UTC 0?)
> 
> Should the stream also be aware of this timezone? How does cmp between timestamp with timezone and timestamp work?
> 
> I guess the stream case does not require timezone if it assumes that the timestamp column and the result of `now()` share the same timezone?

I think our epoch originally from UNIX timestamp is time zone independent (UTC). However, when we are trying to display it to users, they have their own timezone which could be modified by `TimeZone` session variable as well.

**Comment by chenzl25** [comment]
> > Once we intend to support `now()`, we are better to consider it with the timezone. PostgreSQL has a session variable `TimeZone` to allow users to set the time zone for displaying and interpreting time stamps.
> > ```
> > tpch=# select now();
> >               now
> > -------------------------------
> >  2023-01-03 16:24:21.178595+08
> > (1 row)
> > 
> > tpch=# set timezone = 'UTC';
> > SET
> > tpch=# select now();
> >               now
> > -------------------------------
> >  2023-01-03 08:24:48.025885+00
> > (1 row)
> > ```
> 
> Should we use `Timestamptz`?

Yes. If we provide a `now()` function, it will be definitely used by users to insert data to our system. We need to make sure it has a clear definition, e.g. which timezone it belongs to and how we interpret a timestamp.

**Comment by chenzl25** [comment]
Here I give an example I think we need to support. `now()` should return a `timestamp with time zone`, even if we don't allow the user to modify the timezone session variable at this moment.
```sql
create table tz (a timestamp with time zone);
insert into tz values(now());
```

**Comment by soundOfDestiny** [comment]
> > > Once we intend to support `now()`, we are better to consider it with the timezone. PostgreSQL has a session variable `TimeZone` to allow users to set the time zone for displaying and interpreting time stamps.
> > > ```
> > > tpch=# select now();
> > >               now
> > > -------------------------------
> > >  2023-01-03 16:24:21.178595+08
> > > (1 row)
> > > 
> > > tpch=# set timezone = 'UTC';
> > > SET
> > > tpch=# select now();
> > >               now
> > > -------------------------------
> > >  2023-01-03 08:24:48.025885+00
> > > (1 row)
> > > ```
> > 
> > 
> > Should we use `Timestamptz`?
> 
> Yes. If we provide a `now()` function, it will be definitely used by users to insert data to our system. We need to make sure it has a clear definition, e.g. which timezone it belongs to and how we interpret a timestamp.

but `now()` in stream is timestamp without time zone...

**Comment by chenzl25** [comment]
One of the most scenarios of batch `now()` is used to insert the current timestamp (with timezone) into a table. This behavior is also compatible with PostgreSQL.  I think streaming `now()`  can also return a timestamp with UTC timezone. cc @fuyufjh .

**Comment by soundOfDestiny** [comment]
> One of the most scenarios of batch `now()` is used to insert the current timestamp (with timezone) into a table. This behavior is also compatible with PostgreSQL. I think streaming `now()` can also return a timestamp with UTC timezone. cc @fuyufjh .

I have just implemented `Timestamp` `now()` in streaming and `Timestamptz` `now()` in batch. seems we can have this at the moment

**Review by chenzl25 — COMMENTED** [review]
> Here I give an example I think we need to support. `now()` should return a `timestamp with time zone`, even if we don't allow the user to modify the timezone session variable at this moment.
> 
> ```sql
> create table tz (a timestamp with time zone);
> insert into tz values(now());
> ```

Could you please add this case to the e2e test?

**Comment by soundOfDestiny** [comment]
> > Here I give an example I think we need to support. `now()` should return a `timestamp with time zone`, even if we don't allow the user to modify the timezone session variable at this moment.
> > ```sql
> > create table tz (a timestamp with time zone);
> > insert into tz values(now());
> > ```
> 
> Could you please add this case to the e2e test?

done

**Review by chenzl25 — CHANGES_REQUESTED** [review]
Insert into a table column without timezone will cause panic now. 
```sql
create table t (a timestamp);
insert into t values(now());
```

**Comment by soundOfDestiny** [comment]
> Insert into a table column without timezone will cause panic now.
> 
> ```sql
> create table t (a timestamp);
> insert into t values(now());
> ```

Is there any implicit cast which we can utilize?

**Comment by chenzl25** [comment]
> ```sql
> create table t (a timestamp);
> ```

I find a related issue https://github.com/risingwavelabs/risingwave/issues/5826. In our case, it need to convert timestamptz to timestamp.

**Comment by soundOfDestiny** [comment]
> Insert into a table column without timezone will cause panic now.
> 
> ```sql
> create table t (a timestamp);
> insert into t values(now());
> ```

fixed

**Review by chenzl25 — APPROVED** [review]
LGTM, thanks!

**Review by fuyufjh — APPROVED** [review]
LGTM

**Comment by jon-chuang** [comment]
> fixed

I don't think that implicit cast to UTC+0 works here. As pointed out in the tracking issue https://github.com/risingwavelabs/risingwave/issues/5826, we require the implicit timezone to be implemented.

So for batch, we require:

```sql
set time zone = 'UTC+8'
create table t (a timestamp);
insert into t values(now());
select *, now() from t;
```
Result:
```
'2023-01-04 11:25:00' | '2023-01-04 03:25:00Z'
```


> I think streaming now() can also return a timestamp with UTC timezone.

Indeed. It should be a timestamptz not timestamp.

To compare against timezone, performed by the MV should cast timestamp -> timestamptz with the session timezone.

So for instance:
```sql
set timezone = 'UTC+8';
create table t (ts timestamp);
select * from t where ts >= now();
```

plan pseudocode:
```
DynamicFilter { ts >= now() }
  Project { ts::timestamp with time zone("UTC+8") }
  Now { now (UTC) }
```

In order to more finely control the timezone of their timestamp column, the user should cast their timestamp to a specific timezone.

Data collected in UTC+8, session timezone is UTC-8:
```
create table t (ts timestamp);
select * from t where ts at time zone 'UTC+8' >= now();
```

**Comment by chenzl25** [comment]
This PR doesn't introduce session timezone, so session timezone can be assumed as 'UTC+0' and can't be modified now. Session timezone is important to batch and streaming `now()` and we need to supported it later.

**Comment by soundOfDestiny** [comment]
> MV should cast timestamp -> timestamptz with the session timezone at the time of its creation.

I doubt whether MV should cast timestamp -> timestamptz with the session timezone at the time of its creation.

**Comment by jon-chuang** [comment]
> I doubt whether MV should cast timestamp -> timestamptz with the session timezone at the time of its creation.

It's definitely worth discussing. Note that for batch, psql will cast timestamp to the session timezone. We may forbid such a cast for streaming in order to force the user to explicitly convert their timestamp to a given timezone. I think this semantic may be clearer?

**Review by jon-chuang — APPROVED** [review]
LGTM

**Comment by soundOfDestiny** [comment]
> > fixed
> 
> I don't think that implicit cast to UTC+0 works here. As pointed out in the tracking issue #5826, we require the implicit timezone to be implemented.
> 
> So for batch, we require:
> 
> ```sql
> # defaults to local time zone
> set timezone = 'UTC+8'; 
> create table t (a timestamp);
> insert into t values(now());
> set timezone = 'UTC+0';
> select *, now() from t;
> ```
> 
> Result:
> 
> ```
> '2023-01-04 11:25:00' | '2023-01-04 03:25:00+00'
> ```
> 
> I think we should remove the cast for now.
> 
> > I think streaming now() can also return a timestamp with UTC timezone.
> 
> Indeed. It should be a timestamptz not timestamp.
> 
> To compare against timestamp without time zone, MV should cast timestamp -> timestamptz with the session timezone at the time of its creation.
> 
> So for instance:
> 
> ```sql
> set timezone = 'UTC+8';
> create table t (ts timestamp);
> select * from t where ts >= now() - interval '1 h';
> ```
> 
> plan pseudocode:
> 
> ```
> DynamicFilter { ts >= now() }
>   Project { Cast(ts at time zone 'UTC+8') }
>   Now { now (UTC) }
> ```
> 
> In order to more finely control the timezone of their timestamp column, the user should cast their timestamp to a specific timezone.
> 
> Data collected in UTC+8, session timezone is UTC-8:
> 
> ```sql
> create table t (ts timestamp);
> select * from t where ts at time zone 'UTC+8' >= now() - interval '1 h';
> ```

ok. Now I switch streaming now into `timestamptz` too.
Should I revert the cast from timestamptz to timestamp? @chenzl25

**Pull request merged**